src/liballoc/collections/btree/node.rs

```diff
@@ -492,16 +499,19 @@ impl<'a, K, V, Type> NodeRef<marker::Mut<'a>, K, V, Type> {
     }
 
     /// Returns a raw ptr to avoid asserting exclusive access to the entire node.
+    /// Caller must ensure that this is indeed a proper leaf node, and not the shared root.
     fn as_leaf_mut(&mut self) -> *mut LeafNode<K, V> {
-        // We are mutable, so we cannot be the shared root, so accessing this as a leaf is okay.
+        debug_assert!(!self.is_shared_root());
         self.node.as_ptr()
     }
 
     fn keys_mut(&mut self) -> &mut [K] {
+        debug_assert!(!self.is_shared_root());
         unsafe { self.reborrow_mut().into_key_slice_mut() }
```

Comment:
`into_key_slice_mut` already asserts the same thing... not sure what our usual stanza on redundant debug assertions is, but this seems a little excessive, maybe? If the purpose is just documentation, then IMO that would be better served by a doc comment than a redundant assertion.

Reply:
It does seem a little excessive for a one-line function, but I can see advantages to having both. I don't trust comments much, if they're clear at all, like that "Assert that this is indeed a proper leaf node": it took me weeks to understand this was a request towards callers.

What's your (or the stanza's) opinion on where the actual assertion should be: the external entry point(s), or close to where the damage would happen?

Reply:
> "Assert that this is indeed a proper leaf node"

Yeah that's a bad comment. That's why I suggested edits in this PR before to say "the caller has to ensure that...".

> What's your (or the stanza's) opinion on where the actual assertion should be: the external entry point(s), or close to where the damage would happen?

My gut feeling is, close to the damage.